grr/server/grr_response_server/flows/general/osquery_test.py

```diff
@@ -147,48 +152,108 @@ def testFailure(self):
       self._RunQuery("UPDATE time SET day = -1;")
 
 
-class FakeOsqueryFlowTest(flow_test_lib.FlowTestsBaseclass):
+def _CreateDictTable(column_number, row_number) -> List[OrderedDict]:
+  def CellValue(col_number, row_number):
+    return (f"col-{col_number}", f"col-{col_number}, row-{row_number}")
+
+  return [
+    OrderedDict(
+      CellValue(col_number, row_number) for col_number in range(column_number)
+    ) for row_number in range(row_number)
+  ]
 
+
+class FakeOsqueryFlowTest(flow_test_lib.FlowTestsBaseclass):
   def setUp(self):
     super(FakeOsqueryFlowTest, self).setUp()
     self.client_id = self.SetupClient(0)
 
-  def _RunQuery(self, query):
+  def _NewTestFlowStartAndRun(self, query):
     session_id = flow_test_lib.TestFlowHelper(
         osquery_flow.OsqueryFlow.__name__,
         action_mocks.ActionMock(osquery_action.Osquery),
         client_id=self.client_id,
         token=self.token,
         query=query)
+    return session_id
+
+  def _RunQuery(self, query):
+    session_id = self._NewTestFlowStartAndRun(query)
     return flow_test_lib.GetFlowResults(self.client_id, session_id)
 
+  def _AssertTablesMatch(
+    self,
+    table: rdf_osquery.OsqueryTable,
+    expected: List[OrderedDict]
+  ) -> None:
```

Comment:
Ultimately I think it's a trade-off, and it's a matter of taste which one is better. 
Benefits of "AssertTablesMatch":
1. To me the test bodies are more readable with "AssertTablesMatch" as opposed to having a few assertions. 
2. If a test fails, the stack trace will include "AssertTablesMatch", so it is obvious that the problem is in the tables. Whereas with plain assertions we get no semantic context.

Downsides:
1. If a test fails, you need to look into the implementation of AssertTablesMatch, and figure out what is failing.
I'm biased but it's not difficult to decode `self.assertEqual(table_col.name, expected_col_name)` or `self.assertEqual(table_col_values, expected_col_vlues)`. (Jokes on me, I have a mistype)

I don't feel too strongly about it, though, so I can revert your code to how it used to be.

For mine, however, I will need a minimum of 10 additional assertions in the non-truncating test, and 11 in the other one. And what if we decide to increase the truncating length to 20? I need the automation.

What are your thoughts?

Reply:
> To me the test bodies are more readable with "AssertTablesMatch" as opposed to having a few assertions.

I would not count it as a benefit, because it is not objective. For me it is less readable and forces me to execute code in my head and jump around the code instead of reading in in a linear way.

> If a test fails, the stack trace will include "AssertTablesMatch", so it is obvious that the problem is in the tables. Whereas with plain assertions we get no semantic context.

The name of the test case should be everything that you need. You are testing tables, so obviously the problem will be with the tables. If you test more then one thing, then your test is wrong and should be split.

> I don't feel too strongly about it, though, so I can revert your code to how it used to be.

I kind of to, so I would prefer to revert to the old code.

> For mine, however, I will need a minimum of 10 additional assertions in the non-truncating test, and 11 in the other one

10 or 11 is within reason, IMO.

>  And what if we decide to increase the truncating length to 20? I need the automation.

That's kind of the problem here that you depend on an implementation detail. Truncating length should be a parameter and in the tests you should cover two scenarios:

  * Use some small truncating length to write tests with particular values that are asserted.
  * Use some big truncating length to assert have some vague assertion that the "shape" is still preserved.